sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/udaf.scala

```diff
@@ -450,3 +452,165 @@ case class ScalaUDAF(
 
   override def nodeName: String = udaf.getClass.getSimpleName
 }
+
+/**
+ * The internal wrapper used to hook a [[UserDefinedImperativeAggregator]] `udia` in the
+ * internal aggregation code path.
+ */
+case class ScalaUDIA[T](
+    children: Seq[Expression],
+    udia: UserDefinedImperativeAggregator[T],
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0)
+  extends TypedImperativeAggregate[T]
+  with NonSQLExpression
+  with UserDefinedExpression
+  with ImplicitCastInputTypes
+  with Logging {
+
+  def dataType: DataType = udia.resultType
+
+  val inputTypes: Seq[DataType] = udia.inputSchema.map(_.dataType)
```

Comment:
Are inputs assumed to be nullable?

Reply:
Inputs are not assumed to be nullable, however the logic around that is implicitly embodied inside the `update` method: the implementation of `update` can either check for null inputs or not.  It might be possible to define a nullable-update and non-nullable-update, so that checks for null input can be skipped if the aggregation is on some columns known to never contain null values.  The nullable-update could default to wrapper around the non-nullable-update that skips if any input is null, with the option to override. Encoders support this kind of dual functionality, for example.  

Based on my performance studies, my intuition is that checks for null inputs are not likely to add significant overhead relative to all the other compute going on, however if the common use-case could ignore that definition maybe it's worth allowing the option

Reply:
As you said, I think this is important for performance. It also affects the output type. One valid way to handle null input values is to make the output null. For example, `avg(1, null, 3)` is `null`. If input values are guaranteed to be non-null, then functions like this would guarantee a non-null result and wouldn't need to check for null values.

I think it's important that this be a full struct type, not just the individual data types.